ipcache: Add ability to override identity via UpsertMetadata

gandro opened the pull request:
The new asynchronous IPCache API tries to determine the identity of a
prefix based on the provided labels. However, in some cases, the sources
providing the IPCache metadata (e.g. CiliumNode, CiliumEndpoint) already
know the numeric identity of the prefix. In that case, we want to give
that predefined global identity preference over any local identity we
might derive from the labels.

All other metadata is still tracked. If the override is ever removed
from the prefix, the metadata will be used to determine the new prefix
identity.

Currently, the override identity is always accepted, regardless whether
the numeric identity is known e.g. to the identity allocator. This
behavior mirrors the current semantics of the legacy IPCache.Upsert
method. In the future, we could consider blocking overrides until they
are actually known to the identity allocator.

Identities added via override do not trigger a policy map update. It is
assumed that the policy map update for those identities is handled
elsewhere. This again mirrors the current behavior of IPCache.Upsert.

joestringer committed:
ipcache: Consolidate identity resolution

An upcoming commit will introduce a way to resolve an identity in the
prefixInfo to a previously-allocated identity, rather than the identity
that this logic may allocate. In order to simplify upcoming work to
resolve the identity from a prefixInfo, this commit refactors the
callers into a single function with all of the corresponding logic,
resolveIdentity().

The prior comments about conflicts between ToCIDR policy and the
identity resolution logic here for the metadata is still valid for now.
This is because ToCIDR policy does not yet use the new IPCache metadata
APIs that call into these functions, so there is no way that this logic
could resolve that conflict. That said, the new core structure of the
IPCache.metadata is designed to handle that conflict, and that will not
require any subsequent changes to this function. So I just dropped the
comments / references to GH-17962.

This commit now also documents the datapath assumption being tied to the
remote-node logic here, so that if we do end up providing a way to
associate additional remote-node labels with IPs corresponding to remote
nodes, hopefully the future contributor will be able to reason about the
impact and prevent weird bugs / interactions with that new logic. This
assumption already existed, it is not introduced by this commit.

There may be a minor change in the behaviour now, in that if a caller
somehow hands the IP of the host or a remote node into AllocateCIDRs()
then previously the function would erroneously allocate a new CIDR
identity for those IPs. The function will now consistently associate the
reserved identity for those IPs.

Other than that, this commit is not expected to have any side effects.

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer committed:
ipcache: Document remote-node identity assumption

As described in the comment added by this commit message, this logic is
making assumptions about the way that identities for remote nodes are
allocated, which hold true today but may not hold true in the future. If
we do end up providing a way to associate additional remote-node labels
with IPs corresponding to remote nodes, hopefully the future contributor
will be able to reason about the impact and prevent weird bugs /
interactions using this new comment.

Signed-off-by: Joe Stringer <joe@cilium.io>

gandro committed:
ipcache: Add ability to override identity via UpsertIdentity

The new asynchronous IPCache API tries to determine the identity of a
prefix based on the provided labels. However, in some cases, the sources
providing the IPCache metadata (e.g. CiliumNode, CiliumEndpoint) already
know the numeric identity of the prefix. In that case, we want to give
that predefined global identity preference over any local identity we
might derive from the labels.

All other metadata is still tracked. If the override is ever removed
from the prefix, the metadata will be used to determine the new prefix
identity.

Prefixes added via UpsertIdentity otherwise are treated similarly to
ones upserted via UpsertMetadata: The policy maps are still updated for
any newly added identity, and the identity is removed from policy maps
if RemoveIdentity caused the identity to be deallocated.

Co-authored-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
testidentity: Clean up associated labels after release

This fixes a nil pointer access in AllocateIdentity, where a new
identity with the identical label set of a previously deleted identity
which would try to bump the reference count of the deleted identity.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro commented on pkg/ipcache/metadata.go:
```diff
@@ -221,36 +246,58 @@ func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.P
-			// releasing the previous reference.
-			if _, ok := idsToAdd[id.ID]; !ok {
-				previouslyAllocatedIdentities[prefix] = id
+			_, oldIdentityReallocated := idsToAdd[oldID.ID]
```
I tried to keep the old logic here, but I am not 100% certain why we're checking `oldID` here, and not `newID`. If both are the same (e.g. if reached via the following statement), then this ensures that we're not releasing any identity that we've just allocated.

https://github.com/cilium/cilium/blob/4bafe1750f7b67fcab4b057c6e1aa9e8b9142584/pkg/ipcache/metadata.go#L224-L226

But if `oldID != newID`, then I don't know what this is intending to do, since even if oldID was reallocated I think we still want to call release on `oldID` to ensure proper balance between allocate/release.

joestringer replied:
This is tricky. It may turn out that the mechanism we're using here is more complicated than it needs to be. I'll mull over some thoughts below.

One thing that tripped me up here was the notion of "reallocating an identity", as I'm not entirely sure the intent of what that's supposed to mean. Does that mean that the same identity got allocated again (=> increment the refcount, which must be eventually decremented)? Or we decided to change the labels for this prefix, and hence "re-"allocate a new identity for the prefix? (=> need to decrement the old identity reference to potentially release that entry)

I think there's maybe three cases that hit this path...?
1. The old identity and new identity is the same for this prefix. In this case, there's a `goto` shortcut above that jumps to here without populating `idsToAdd[oldID.ID]`. The lookup will return a false boolean result, so we will release the additional reference incremented by the `injectLabels()` call. 
2. The old identity and the new identity is different for this prefix. In this case, `idsToAdd[newID.ID]` was inserted so then if we look up the old identity in this map then it should return false and that is a sign that we need to release the old identity.
3. The prefix no longer has any information associated, so we are trying to release references to the identities + delete the entry. Then a lookup will return false and that's the sign to release the identity below.

So I think that the conditional check here was probably not actually necessary before? We always have one spare reference to the identity from the logic above that we need to release?

That said, `idsToAdd` is still necessary for another different purpose, which is to propagate updates into the selectorcache correctly.

gandro replied:
Thanks a lot for your thoughts on this, that list of cases was super helpful. 

> So I think that the conditional check here was probably not actually necessary before?

The only case where the check would be true is if another prefix newly allocated oldID and our prefix is no longer using oldID. In that case however, I think we still would need to actually release the identity, no? Otherwise our previous allocation call is never matched with a release. In other words, this could be a sequence of events with the condition in place:

> 
> _Metadata:_
> Prefix 1.1.1.1 has label foo.
> 
> _First round of `InjectLabels` is triggered:_
> 1. prefix 1.1.1.1 allocates identity X. `idsToAdd[X] = {foo}` is set. No previous entry exists for the prefix, so no release is happening
> 
> _Updated Metadata:_
> Prefix 1.1.1.1 has label bar (previous label foo was removed)
> Prefix 2.2.2.2 has label foo
> 
> _Second round of `InjectLabels` is triggered:_
> 1. prefix 2.2.2.2 also allocates identity X, `idsToAdd[X] = {foo}` is set. no previous entry exists for the prefix, so no release is happening. 
> 2. prefix 1.1.1.1 allocates new identity Y. `idsToAdd[Y] = {bar}` is set. A previous entry exists for the prefix, but `idsToAdd[X]` is set, **so no refcount decrease for Y is happening**
> 
> _Updated Metadata:_
> Prefix 1.1.1.1 has label bar (no change)
> Prefix 2.2.2.2 is removed
>
> _Third round of `InjectLabels` is triggered:_
> 1. prefix 2.2.2.2 has a previous entry, but prefixInfo is nil, so the refcount of for identity X is decreased without any allocation of a new identity.
> 
> _Final state:_
> Identity X is not in use by any prefix anymore, but we only had one release calls for two allocs, so it's not released.


Thus I think the condition can cause identity leaks?

joestringer replied:
> The only case where the check would be true is if another prefix newly allocated oldID and our prefix is no longer using oldID. In that case however, I think we still would need to actually release the identity, no? 

Today, this is not possible because the identity for a prefix includes labels derived directly from the prefix. So if the prefix is distinct, the labels are also distinct. Note that if we break this, this would also cause the prefix to be no longer selected by CIDR policy so we have an incentive not to remove this.[^1]

Looking over your logic, I think you're right. That's a great lifecycle example. What's interesting also is that that is dependent on the ordering! Given that the identity release occurs within the same loop iteration, if prefix 1.1.1.1 was updated first then we would release the corresponding reference and then move on to prefix 2.2.2.2 and appropriately allocate a new reference. In the end, this would balance the refcount adds/deletes. However, in the order you provided, it would have this issue.

I think it's fair to say that we haven't previously reasoned through this core loop with the idea in mind that multiple ipcache entries could share the same identity. In the end, the intent is that we either insert a new ipcache entry and then rely on the deletion of that entry to release the reference, or we should release the additional reference here in the loop. This logic that relies on the map from potentially other iterations is not accurately achieving this goal.

[^1]:　Not a hard requirement, technically if we have no CIDR policy then we don't need to put CIDR labels into the identity but this is how we do it today. This way, if a CIDR policy is added later, we don't have to re-evaluate the identities for all existing CIDR identities.

gandro replied:
> So if the prefix is distinct, the labels are also distinct.

Ahh! The fact that CIDR identites are different for each prefix is a piece I missed. I think we're also deriving identity 6 and 7 from the labels at the moment - those are shared, but they are also static, so alloc/release are no-ops afaiu.

So In conclusion we feel fairly confident that we can actually remove the `_, ok := idsToAdd[id.ID]; !ok` condition, as never seems to be `false`?

joestringer replied:
So far we've explored the options:
- same as before
- different from before
- removed
- conflicts between iterations of the loop

The one case we didn't consider is:
- add a new identity

However, for this case we already check `if entryExists { ... }` so we would skip the release of the old (non-existing) identity.

The other angle to think about this is, which identity might we *want* to add a reference on? And the answer to that is only the "new" identity - ie the identity corresponding to the prefix that we intend to push into the bpf ipcache map which is not already there. 

So yes, I think we can unconditionally release that reference to the old identity. There's no need for this extra map lookup call.

gandro replied:
Removed the condition.

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -67,6 +67,14 @@ type metadata struct {
+	// overriddenIdentitiesMu protects overriddenIdentities and must be taken
+	// both before the metdata.RWMutex or the IPCache.Lock
+	overriddenIdentitiesMu lock.Mutex
+	overriddenIdentities   map[netip.Prefix]struct{}
```
:bulb: 

Yeah the reference tracking is the trickiest part of this API. Broadly we have two alternatives I can think of, and it looks like we've explored both in our two PRs looking at solving this problem :-)

Here I recognize the more explicit form of having a separate map to track the prefix to avoid adding/removing reference counts for the identities that are provided explicitly. This has a couple of nice properties: it's less "magic", and there's no additional messing around with the reference counts.

The alternative is that the core logic takes in an identity, then internally adds its own reference counts to the identity. I had explored that option briefly here: https://github.com/cilium/cilium/pull/21675/commits/2d2bee88f7f095fa1331c272b93c94c9796eace4 , but based on the TODO comments I have in that commit it seems like I was already leaning towards an approach more like this.

gandro replied:
I am unfortunately not super familiar with identity allocation. I chose the separate map because it more closely mirrors the existing behavior. If I understand your implementation correctly, it's main idea is to determine the identity based on the labels instead of the numeric identity, correct?

My main worry here was that for CiliumNode objects (which can get assigned an override identity via CEW), we wouldn't have the labels available that lead to the identity being assigned to the node. But re-reading the code, it seems that the labels from the CEW are written into the CiliumNode object together with the assigned identity, so that should work too.

The other open question I have is around the policy map update. Here I'm out of my depth if it makes sense to update the policy maps for "override identities", which I assume are already added to the policy map via policy and if this is something we need to track here, or if it is something that is anyway already tracked by the selector cache.

joestringer replied:
Yeah pretty much as you say. The idea is that we'll pull the labels out from the identity and then allocate the identity for that set of labels. The idea behind that is that the identity allocator would figure out that this is an existing identity and increment the refcount on that identity + return the identity that was already existing.[^1]

Technically the "override identities" do not need to trigger policy update in this logic, correct. This is where the role of IPCache in the overall architecture is changing a bit in general - The new model says that IPCache itself is responsible for generating new identities (in the other case), and hence it must notify the policy engine for those new identities so that the policymaps are appropriately plumbed to handle this logic.

For "override identities", the corresponding logic for policy integration is currently elsewhere. We could potentially decide that actually, rather than having that handled elsewhere, instead we rely on this notification from the ipcache. At a high level, this is saying that when some watcher or other agent logic learns that "IP X has identity Y", it informs the userspace IP informationbase ("IPCache") about this mapping, then IPCache is responsible for ensuring that all relevant calculation occurs, and occurs in the right order - that is to say, inform the policy engine + update bpf policy maps and then update the bpf ipcache map. I think this is a reasonable design and could potentially simplify some other logic. 

That said, for now it's not necessary. I think your question is coming from the perspective "does it need to notify the policy engine, and if not, should we skip it?" which is quite a reasonable question to ask.. technically we could skip it for now. I think that'll make the logic slightly more complex. Would it hurt to notify the policy engine twice for the same identity updates? Well, that's a bit more tricky. :thinking: At a glance it looks like `SelectorCache.UpdateIdentities()` is idempotent so in a sense it should be OK. However, if an identity is added + deleted within a quick succession, and there are two different pieces of agent logic notifying the selectorcache at different times outside of a critical section then we could end up with inconsistent state in the selectorcache. So from that perspective it should probably be the responsibility of exactly one piece of code to notify the selectorcache about the identity update. We can ensure consistency by queuing events in order as they come in (which we already do).

^^ Part of the idea behind having a dedicated goroutine handling the ipcache updates like this is to force serialization of such events so that we don't end up in some intermediate inconsistent state in the selectorcache.

[^1]: Then we still have the question of how that reference is eventually removed, which I covered in a bit more detail in one of the other discussion threads below. The principle is that this refcount would stay live as long as the corresponding ipcache entry stays live; if the ipcache entry already exists and we re-allocate the same identity then we need to make sure that this doesn't cause an identity reference leak.

gandro replied:
In order to keep the code simpler, I decided to not treat override identities special when it comes to notifying the policy engine and have InjectLabels do the notification for all identities. Let's see how that works out.

joestringer commented on pkg/ipcache/types.go:
```diff
@@ -91,3 +107,30 @@ func (s prefixInfo) Source() source.Source {
+					logfields.Resource:            idResource,
+					logfields.ConflictingIdentity: v.identityOverride,
+					logfields.ConflictingResource: resource,
+				}).Warning("Detected conflicting identity override for prefix")
```
Probably should detect+log this during insert rather than get?

gandro replied:
Agreed!

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -273,16 +320,28 @@ func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.P
 		}
 	}
 
+	ipc.metadata.overriddenIdentitiesMu.Unlock()
```
There's one slightly odd thing about the locking here compared to how the documentation specifies: `ipc.mutex` will be locked while holding `overriddenIdentitiesMu` but released after releasing `overriddenIdentitiesMu`. Not sure whether there's requirement for the ipcache lock to be purely nested inside but I thought I'd highlight it for discussion either way.

gandro replied:
Good point. We could consider moving `ipc.metadata.RUnlock` further down. This is also another downside of this approach: IPCache now has three locks where we need to take care of ordering, with further complicates the logic.

joestringer replied:
I had wondered about that, but that also made me a bit nervous since we've been dealing with deadlocks between policy generation and ipcache updates recently, and I'm not sure whether it will be safe to hold the metadata lock across the UpdatePolicyMaps call. I think it's slightly safer now given endpoint policy generation doesn't synchronously release ipcache state (https://github.com/cilium/cilium/pull/21629), but this stuff is notoriously tricky to understand the implications. :/ 

On the plus side, we can get something like this into the master branch and get some sort of testing + cooking as part of the dev&release process.

joestringer replied:
^ Thinking a bit more, metadata lock is grabbed by all sorts of different pieces of code that injects information into the IPCache. Those callers should not be blocked by the IPCache waiting on the policy updates to complete. 

Was there something in particular that was preventing us from just deferring the lock release in this function, other than just trying to minimize the critical section as much as possible?

gandro replied:
> Was there something in particular that was preventing us from just deferring the lock release in this function, other than just trying to minimize the critical section as much as possible?

Not as far as I can tell.

The new version of the PR doesn't need to change any of the locking behavior anymore, so marking this as resolved for now.

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -221,36 +246,58 @@ func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.P
+
+			if !override {
+				// We allocated a new identity, which means we want to add it
+				// to the policy map
```
I guess we'll need to document the inverse of this in the UpsertIdentity() API: Callers from policy context must arrange for policymaps to be updated themselves.

gandro replied:
Absolutely agree. At the moment, there is no separate method to inject override identities, but it might make sense to have one even just for documentation purposes.

joestringer reviewed (commented):
I would _like_ to not have to complicate that core loop further by adding another fork in the logic, but at the same time I'm not sure whether we really have another good option.

Right now, the logic is looking correct to me. It wouldn't hurt to stew it over another day and take another fresh look but overall this PR seems correct as-is, modulo one minor locking quirk.

I also posted https://github.com/cilium/cilium/pull/21675 for a comparison of a similar approach, but I think that your PR's handling of the identity reference counting is more mature than #21675. Please take a look either way, in case there's any useful insights or approaches there that could improve this PR.

I added a few discussion threads below but having spent a chunk of time with this one I'd be OK with dropping most of my one + rebase any other remaining bits on top of this.

gandro committed:
ipcache: Track if an entry was created via InjectLabels

This commit adds a new flag to each IPCache entry, tracking if it was
upserted using the new asynchronous metadata infrastructure or the
legacy Upsert/Delete API.

This is needed to properly track if an entry can safely be deleted once
no metadata is associated with it anymore. Previously, the code relied
on the identity reference count on this, but that does not work for
reserved identities, or non-CIDR identities which may be shared among
multiple prefixes. The latter was not a concern so far, but will be
added in a future commit. For the former (reserved identities), the
existing code relied on mixed API use for each prefix, requiring that
other code eventually invoked a manual Delete call on each entry.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
testidentity: Increase reference count for global identities too

Previously, the mock identity allocator would allocate a new identity
for each global labelset. This is not how the real allocator behaves.
This commit therefore changes the behavior of the mock allocator to
return the already existing identity and bumping its reference count.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Rename local variable name

This renames the `id` variable to `oldID`, to better distinguish it from
the `newID` variable and indicate that `oldID` is the one about to be
replaced or deleted.

This commit contains no functional changes.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Remove superfluous if condition

See https://github.com/cilium/cilium/pull/21667#discussion_r992445940
for an in-depth discussion about this change. For local identities, this
condition never evaluated to false, since identities are different for
each prefix and thus no other prefix could have added a map entry for
the previously allocated identity.

A subsequent commit will cause InjectLabels to also allocate global
identities. In that case, the condition is actively harmful, since it
could cause an identity leak.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

joestringer commented on pkg/testutils/identity/allocator.go:
```diff
@@ -70,7 +70,7 @@ func (f *MockIdentityAllocator) AllocateIdentity(_ context.Context, lbls labels.
 	requiresGlobal := identity.RequiresGlobalIdentity(lbls)
 
-	if numID, ok := f.labelsToIdentity[lbls.String()]; ok && !requiresGlobal {
+	if numID, ok := f.labelsToIdentity[lbls.String()]; ok {
```
:+1: this change is good. It may even help with detecting global identity leaks with the new interface. I didn't get to reviewing the commit where that can occur yet so maybe you were already thinking this but if not, this may be something to explore to verify the core refcounting logic with these override identities.

joestringer replied:
^^ After reviewing the tests, it looks like identity reference counting is not validated by the current logic. It'd be nice to include that just to verify that the core logic isn't leaking additional references.

gandro replied:
The `TestOverrideIdentities` does indirectly validate that reference count is bumped, by checking that the numeric identity added to the IPCache is equal to the one allocated before  `UpsertIdentity` is called:

https://github.com/cilium/cilium/blob/b95b85726b1cdf8130fa6216075ae1bce00f1a74/pkg/ipcache/metadata_test.go#L159-L161

But yeah, I agree that having a more explicit test would be nice. Will add.

gandro replied:
Added ref count assertions.

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -482,6 +510,27 @@ func (ipc *IPCache) RemoveLabels(cidr netip.Prefix, lbls labels.Labels, resource
+// lbls set, a warning is emitted and the numeric identity here is ignored.
+// The given labels are not modified nor is its reference saved, as they're
+// copied when inserting into the map.
+// This will trigger asynchronous calculation of any local identity changes
+// that must occur to associate the specified labels with the prefix, and push
+// any datapath updates necessary to implement the logic associated with the
+// metadata currently associated with the 'prefix'.
```
Given that the caller presumably holds a reference to the target (numeric) identity, but then this interface will asynchronously perform the updates itself, does the caller need to retain a reference until the underlying async logic runs? Or will this implementation grab a preliminary reference on the corresponding identity up until the async logic is run?

The concern I have is something like, the caller does:
- Allocate identity
- UpsertIdentity(...)
- Release reference identity

Then, since `UpsertIdentity(...)` internally defers the insert logic, that could happen after the caller releases _its_ reference on the identity, and then it becomes more likely that the ipcache will (re-)allocate a new numeric identity for these labels. This is not ideal.

joestringer replied:
For what it's worth, I think that the current code does handle this case gracefully, it will just log an extraneous warning that the user can't do much about. And I guess at least right now, there are no callers of this interface so there's no way to trigger this condition so it's not exactly a big deal. But it would be nice to make this expectation super clear for future callers.

If we used this from the endpoint logic, then the endpoint itself would already retain a reference to the identity so I think that case would be fine. Not sure about other users that directly allocate an identity.

gandro replied:
Yeah, I was designing this interface originally with `NodeManager` in mind, which actually does not even locally allocate the identity it's passing in here, it just blindly trusts the remote node that it holds a reference to the identity it is claiming to have. So the code as I wrote it doesn't really care if the identity already exists or not, the numeric identity right now is only there to provide a warning, it doesn't have any functional purpose. 

I'm myself a bit partial to the requirement that the caller of `UpsertIdentity` needs to provide a numeric identity (which implies prior allocation). Maybe it would be cleaner for `UpsertIdentity` to just require the labels as your PR originally did (and _maybe_ have the numeric identity as an optional argument just for logging/troubleshooting purposes)

---

In the end, I guess we do assume that if the caller of `UpsertIdentity` needs to somehow know the numeric identity of a prefix, then we also assume that someone else is already holding the reference to it. 

I guess in theory it could happen that code allocates an identity, schedules `UpsertIdentity`, then drops the identity again before `InjectLabels` (executed asynchronously) had a chance to run. That would re-assign the numeric identity. If that code assumes `UpsertIdentity` instantly keeps the reference/numeric identity alive, then yeah, that would likely be buggy behavior. On the other hand, decreasing an identity ref count while assuming the numeric identity remains valid is dangerous regardless of this API.

gandro replied:
I've removed the numeric identity. That way, the interface is agnostic to the fact if the identity was already allocated or not.

I've also renamed the method to `OverrideIdentity` to distinguish it a bit more clearly from `UpsertLabels`.

joestringer commented on pkg/ipcache/types/types.go:
```diff
@@ -55,3 +56,19 @@ func NewResourceID(kind ResourceKind, namespace, name string) ResourceID {
+// Must be provided together with a set of labels. Any other labels associated
+// with this prefix are ignored while an override is present.
+// This type implements ipcache.IPMetadata
+type IdentityOverride identity.NumericIdentity
```
nit: the other types in this package are all used outside the `pkg/ipcache` package, vs.  `pkg/ipcache/types.go` which includes ipcache-internal types. We could probably keep this in the internal-only package given this is only used inside `pkg/ipcache/` for now, plus we already have to update `pkg/ipcache/types.go` to handle this type.

joestringer commented on pkg/ipcache/types.go:
```diff
@@ -15,13 +18,36 @@ import (
+					logfields.Resource:            overrideResourceID,
+					logfields.ConflictingIdentity: v.identityOverride,
+					logfields.ConflictingResource: resourceID,
+				}).Warning("Detected conflicting identity override for prefix")
```
Suggestion for a bit more clear user impact:

```suggestion
				}).Warning("Detected conflicting identity override for prefix. This may cause connectivity issues for this address.")
```

It'd be nice to include the prefix in the logs here too, to assist the user in understanding the potential impact.

gandro replied:
The prefix is already included in the logs, check the call site:

https://github.com/cilium/cilium/blob/efc9baae0677fe37166172b6885cc6d32813e177/pkg/ipcache/metadata.go#L123

I found that a bit more cleaner than passing in the prefix.

joestringer commented on pkg/ipcache/types.go:
```diff
@@ -91,3 +127,15 @@ func (s prefixInfo) Source() source.Source {
+func (s prefixInfo) IdentityOverride() (types.IdentityOverride, labels.Labels) {
+	for _, v := range s {
+		if v.identityOverride.IsValid() {
+			return v.identityOverride, v.labels
```
This is probably OK in terms of resolving the conflict. IIRC `range s` won't guarantee the same item is picked if there are conflicts and the function is run multiple times. A more stable behaviour would be to pick out the conflicts, sort them, and pick an identity based on some stable heuristic.... but ideally the user will never hit conflicts because we recognize the potential for conflict ahead of time and put guard rails in place. So probably fine to just leave this for now, especially since there are warnings when conflicts are inserted.

gandro replied:
Yeah, I think I mentioned somewhere in the doc to `UpsertIdentity` that a random one will be chosen upon conflict, which is due to this `range s` here being non-deterministic. 

Looking at this again, I am afraid that having a random pick here this will likely just cause headaches if we ever need to troubleshoot conflicting entries. I think I'll change this to use a deterministic iteration order (since follow-up PRs for hostIP and encryptKey will also rely on the the same conflict resolution logic)

gandro replied:
I added basic deterministic conflict resolution by sorting alphabetically. I'm honestly not sure what heuristic would be more useful in practice? Prefer CIDR identities over global ones over reserved ones? For two identities in the same category, which one should we pick?

Because I couldn't come up with a good rationale for anything, I decided to keep it simple for now and just sort alphabetically.

joestringer reviewed (approved):
LGTM. Worth looking a bit closer at some of the questions I have below around semantics for the callers. These are not critical right now since there's no users of the new interface yet, but it won't be long before we'd need to grapple with some of these questions.

tklauser reviewed (approved):
Looks good for @cilium/cli
